perf(core): add option to remove blank text nodes from compiled templates

pkozlowski-opensource opened the pull request:
Closes #11679

pkozlowski-opensource committed:
perf(core): add option to remove blank text nodes from compiled templates

vicb commented on packages/compiler/src/ml_parser/html_parser.ts:
```diff
@@ -21,6 +22,9 @@ export class HtmlParser extends Parser {
-    return super.parse(source, url, parseExpansionForms, interpolationConfig);
+    const parseTreeResult = super.parse(source, url, parseExpansionForms, interpolationConfig);
+
+    return new ParseTreeResult(
```
Might be better not to remove WS here but when the result is used. (wrt i18n)

pkozlowski-opensource replied:
Do you mean a place where a template parser gets HTML parser output (ex.: https://github.com/angular/angular/blob/d71ae278eff8abcc5a7cf58ca7c6c56da66bb35b/packages/compiler/src/template_parser/template_parser.ts#L143-L144)?

I would much prefer removing WS on the HTML parser level as then we don't have to worry about the interpolation case (see https://github.com/angular/angular/pull/18401#discussion_r130311400)

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,54 @@
+export const PRESERVE_WS_ATTR_NAME = 'ngPreserveWhitespaces';
+
+function hasPreserveWhitespacesAttr(attrs: html.Attribute[]): boolean {
+  for (let attr of attrs) {
```
`return attrs.some(...)`

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,54 @@
+  return false;
+}
+
+class WhitespaceVisitor implements html.Visitor {
```
extends existing `RecursiveVisitor`

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,54 @@
+      return new html.Element(
+          element.name, html.visitAll(this, element.attrs), element.children, element.sourceSpan,
+          element.startSourceSpan, element.endSourceSpan);
+    } else {
```
- no `else` after return
- `super.visitElement(...)`

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  visitAttribute(attribute: html.Attribute, context: any): any {
+    return attribute.name !== PRESERVE_WS_ATTR_NAME ? attribute : null;
```
unused - could be dropped once you extend

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  visitText(text: html.Text, context: any): any {
+    return text.value.trim().length > 0 ? text : null;
```
Note:
- `<tag>  *  </tag>` would stay unmodified
- `<tag>  {{ * }}  </tag>` would be changed to `<tag>{{ * }}</tag>` which is wrong `{{ `hello` }} {{ `World` }}`

vicb replied:
Add test

pkozlowski-opensource replied:
Added the following tests:

```javascript
it('should preserve whitespaces between interpolations', () => {
          expect(humanizeDom(parser.parse(`{{fooExp}}\n{{barExp}}`, 'TestComp'))).toEqual([
            [html.Text, '{{fooExp}}\n{{barExp}}', 0],
          ]);
        });

        it('should preserve whitespaces around interpolations', () => {
          expect(humanizeDom(parser.parse(`\t{{exp}}\n`, 'TestComp'))).toEqual([
            [html.Text, '\t{{exp}}\n', 0],
          ]);
        });
```

And both tests are passing as on the HTML parser level we still don't "understand" interpolations (that is, a visitor will see a text node with content like `{{fooExp}}
{{barExp}}`. So if we keep WS removal on the HTML parser level we are good.

vicb commented on packages/compiler/test/ml_parser/html_parser_spec.ts:
```diff
@@ -300,6 +299,61 @@ export function main() {
+
+        it('should remove whitespaces from child text nodes', () => {
+          expect(humanizeDom(parser.parse('<div><span> </span></div>', 'TestComp'))).toEqual([
+            [html.Element, 'div', 0], [html.Element, 'span', 1]
```
please add trailing commit to make format more readable.

vicb reviewed (changes requested):
- Fix the issue with interpolations, add tests,
- Move the WS removals out of the HTML parser (to the `TemplateParser`)

pkozlowski-opensource commented:
@vicb I've addressed all the style comments and pushed a new version. Regarding remaining points:

> Fix the issue with interpolations, add tests,

I've added [tests](https://github.com/angular/angular/pull/18401/commits/e5b203fef0e808086774991f838cefa541d63cc3#diff-220709147438322d4a1f49d09f8a0866R332) but the ones I could think of are passing with the existing implementation (that is, as long as we operate on the HTML AST and not on the template AST). More details: https://github.com/angular/angular/pull/18401#discussion_r130311400

>   Move the WS removals out of the HTML parser (to the TemplateParser)

Somewhere there? https://github.com/angular/angular/pull/18401#discussion_r130331865

mhevery commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,49 @@
+  return attrs.some((attr: html.Attribute) => attr.name === PRESERVE_WS_ATTR_NAME);
+}
+
+class WhitespaceVisitor implements html.Visitor {
```
I don't see a place where we could mark a node as preserve white-space. Something like

```
<div ng-preserve-whitespace>....</div>
````

pkozlowski-opensource replied:
@mhevery it is `<div ngPreserveWhitespaces>...</div>`, see:
* https://github.com/angular/angular/pull/18401/files#diff-220709147438322d4a1f49d09f8a0866R357
* https://github.com/angular/angular/pull/18401/files#diff-fc39bbe62dc11ea559c60cff63c244d1R13

pkozlowski-opensource commented:
> Move the WS removals out of the HTML parser (to the TemplateParser)

Done. @vicb @mhevery this is ready for the review now.

vicb commented on packages/compiler/src/aot/compiler.ts:
```diff
@@ -322,9 +323,11 @@ export class AotCompiler {
         pipe => this._metadataResolver.getPipeSummary(pipe.reference));
 
+    const preserveWhitespaces =
+        normalizePreserveWhitespaceProperty(compMeta !.template !.preserveWhitespaces);
```
can't we store it normalized ?

pkozlowski-opensource replied:
Yeh, it looked wired to me as well. Moved to `DirectiveNormalizer`

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,64 @@
+
+export const PRESERVE_WS_ATTR_NAME = 'ngPreserveWhitespaces';
+
+const SKIP_WS_TRIM_TAGS = new Set(['pre', 'template', 'textarea', 'script', 'style']);
```
option ?

pkozlowski-opensource replied:
Nope. I _was_ sure that I need to skip trimming in `<option>` but a simple test proved me wrong: http://plnkr.co/edit/TqPPra5csHnx2jDYVsNn?p=preview

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,64 @@
+    const isBlank = text.value.trim().length === 0;
+
+    if (!isBlank) {
+      return new html.Text(text.value.replace('&ngsp;', ' ').replace(/\s+/g, ' '), text.sourceSpan);
```
same comment as in the doc wrt `&ngsp;` -> this comes from Angular Dart and should have the same behavior

pkozlowski-opensource replied:
Aligned implementation with the Dart version and added comments.

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,64 @@
+  return attrs.some((attr: html.Attribute) => attr.name === PRESERVE_WS_ATTR_NAME);
+}
+
+export class WhitespaceVisitor implements html.Visitor {
```
Does this need to be exported ?

pkozlowski-opensource replied:
> Please add a comment on the purpose of the class

done

> Does this need to be exported ?

Yes, for tests

vicb replied:
> Yes, for tests

I don't think there is a real need.

vicb commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -345,6 +345,9 @@ class _Tokenizer {
       this._advance();
       const name = this._input.substring(start.offset + 1, this._index - 1);
+      if (name === 'ngsp') {
+        return '&ngsp;';
```
why do we need a special treatment here ? (add a comment)

pkozlowski-opensource replied:
Aligned implementation with the Dart version and added comments.

vicb commented:
Didn't we discuss about a global compiler flag that would be the default (and that we would switch in v6) ?

pkozlowski-opensource commented:
@vicb addressed all the comments from the review, please take another look.

> Didn't we discuss about a global compiler flag that would be the default (and that we would switch in v6) ?

I'm starting to look into it.

vicb commented:
Could you please investigate the failures of the CI for aio, never seen them before (but it has been flaky lately)

vicb commented on packages/compiler/src/compile_metadata.ts:
```diff
@@ -516,7 +519,8 @@ export function createHostComponentMeta(
       externalStylesheets: [],
-      interpolation: null
+      interpolation: null,
+      preserveWhitespaces: false,
```
why false here ?

vicb commented on packages/compiler/src/directive_normalizer.ts:
```diff
@@ -134,6 +142,11 @@ export class DirectiveNormalizer {
       encapsulation = this._config.defaultEncapsulation;
     }
 
+    let preserveWhitespaces = prenormData.preserveWhitespaces;
```
`let preserveWhitespaces = prenormData.preserveWhitespaces !== false;`

Add a comment/factorize the code so that we do not miss updating a location when we release 6.x

vicb commented on packages/compiler/src/directive_normalizer.ts:
```diff
@@ -149,7 +162,8 @@ export class DirectiveNormalizer {
       interpolation: prenormData.interpolation, isInline,
-      externalStylesheets: []
+      externalStylesheets: [],
+      preserveWhitespaces: preserveWhitespaces
```
nit: `preserveWhitespaces `

vicb commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -0,0 +1,78 @@
+
+    if (!isBlank) {
+      // lexer is replacing the &ngsp; pseudo-entity with \uE500
+      return new html.Text(text.value.replace('\uE500', ' ').replace(/\s+/g, ' '), text.sourceSpan);
```
extract `'\uE500'` to a constant

pkozlowski-opensource replied:
Done sth slightly different that should avoid any code duplication / "magic constants" for `\uE500'`

vicb commented on packages/core/src/metadata/directives.ts:
```diff
@@ -675,6 +675,16 @@ export interface Component extends Directive {
+   * as well as speed up components' creation. The whitespace removal algorithm will drop all
+   * the blank text nodes and collapse series of whitespaces to just one space.
+   * Those transformations can potentially influence layout of the generated markup so
+   * the `preserveWhitespaces` should be used with care.
```
Add a plan about the default and plans for v6

pkozlowski-opensource replied:
Done.

vicb commented on packages/compiler/test/template_parser/template_parser_spec.ts:
```diff
@@ -2052,6 +2058,45 @@ The pipe 'test' could not be found ("{{[ERROR ->]a | test}}"): TestComp@0:2`);
+      ]);
+    });
+
+    it('should require explicit whitespace between ICU expansions when removing whitespaces',
```
not should to understand what "should require explicit whitespace between ICU expansions when removing whitespaces" means ? (nor the test)

pkozlowski-opensource replied:
Given `<div>{ count, plural, =0 {small} many {big} } { count, plural, =0 {small} many {big} }</div>` a space between 2 ICU expansions would be removed so we need to explicitly insert `&ngsp;`. This is more of the documentation test, I can remove this one if you think it is not needed.

vicb replied:
- I think the test title (without you extra comment) is misleading - could you change it or add a comment or both ?
- Probably the test should be that it removes spaces between ICU instead ?

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,139 @@
+      it('should replace &ngsp; with a space',
+         () => { expect(visitText('\uE500') !.value).toEqual(' '); });
+
+      it('should treat &ngsp; as a space when collapsing whitespaces',
```
`'should treat &ngsp; as a space when collapsing whitespaces'` this is wrong no ? (a space would lead to `null`)

pkozlowski-opensource replied:
I believe it is correct. Ex. given `<span>&ngsp;\n </span>` we should end up with `<span> </span>` and this is what this test is all about. Please note that `&ngsp;` -> space replacement and trimming happens _after_ checking for a text node being blank (in other words `\uE500` is _not_ considered blank).

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,139 @@
+      });
+
+      it('should remove whitespaces from child text nodes', () => {
+        expect(parseAndRemoveWS('<div><span> </span></div>')).toEqual([
```
Do you have some perf/size figure for text node removal vs WS collapsing only on a real app ?

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,139 @@
+
+      it('should preserve whitespaces between interpolations', () => {
+        expect(parseAndRemoveWS(`{{fooExp}}\n{{barExp}}`)).toEqual([
+          [html.Text, '{{fooExp}} {{barExp}}', 0],
```
why ws instead of `\n` here ? (and below)

pkozlowski-opensource replied:
This is because we are replacing tabs and new lines with spaces to match browsers behaviour. This also makes it easier to trim a sequence of whitespaces to one space.

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,139 @@
+export function main() {
+  describe('whitespaces', () => {
+
+    describe('WhitespaceVisitor', () => {
```
I would not export and test it, testing the function is the same (and reduce the API surface)

pkozlowski-opensource replied:
Done

vicb commented:
Looks very nice, added a few minor comments.

vicb commented on packages/compiler/src/ml_parser/tags.ts:
```diff
@@ -324,4 +329,5 @@ export const NAMED_ENTITIES: {[k: string]: string} = {
   'zeta': '\u03B6',
   'zwj': '\u200D',
   'zwnj': '\u200C',
+  WS_PSEUDO_ENTITY: '\u00A0',
```
confusing ?

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,115 @@
+
+    it('should preserve whitespaces between interpolations', () => {
+      expect(parseAndRemoveWS(`{{fooExp}}\n{{barExp}}`)).toEqual([
+        [html.Text, '{{fooExp}} {{barExp}}', 0],
```
If we can avoid changing the `\n`, it would be preferable as this user might want to have a `\n` here.

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,115 @@
+
+    it('should preserve whitespaces around interpolations', () => {
+      expect(parseAndRemoveWS(`\t{{exp}}\n`)).toEqual([
+        [html.Text, ' {{exp}} ', 0],
```
same here

vicb commented on packages/compiler/test/ml_parser/html_whitespaces_spec.ts:
```diff
@@ -0,0 +1,115 @@
+      ]);
+    });
+
+    it('should ship whitespace trimming in <textarea>', () => {
```
typo: ship
